packages/iTwinUI-react/src/core/Table/actionHandlers/selectHandler.ts

```diff
@@ -56,6 +56,88 @@ export const onSelectHandler = <T extends Record<string, unknown>>(
   onSelect?.(selectedData, newState);
 };
 
+export const onShiftSelectHandler = <T extends Record<string, unknown>>(
+  state: TableState<T>,
+  action: ActionType,
+  instance: TableInstance<T>,
+) => {
+  const resetSelection = () => {
+    instance.rows.forEach((r) => r.toggleRowSelected(false));
+  };
+
+  const getGreaterId = (id: string) => {
+    const greaterId = id.split('.');
+    greaterId[greaterId.length - 1] = `${
+      parseInt(greaterId[greaterId.length - 1]) + 1
+    }`;
+    return greaterId.join('.');
+  };
+
+  const getParentId = (id: string) => {
+    const lastDotIndex = id.lastIndexOf('.');
+    if (lastDotIndex < 0) {
+      return null;
+    }
+    return id.substring(0, lastDotIndex);
+  };
+
+  const isAncestorParent = (currentId: string, endId: string) => {
+    return endId.indexOf(currentId) === 0;
+  };
+
+  // Returns the next row as if it was just a non subrows list
+  const getNextRow = (currentRow: Row<T>, startId: string, endId: string) => {
+    let nextRow;
+    if (
+      (currentRow.subRows ?? []).length > 0 &&
+      isAncestorParent(currentRow.id, endId)
+    ) {
+      // child
+      nextRow = currentRow.subRows[0];
+    } else {
+      // next sibling
+      nextRow = instance.rowsById[getGreaterId(currentRow.id)];
+
+      // If the next sibling doesn't exist, go to the next child at a one upper depth
+      let parentId = getParentId(currentRow.id);
+      while (nextRow == null && parentId != null) {
+        nextRow = instance.rowsById[getGreaterId(parentId)];
+        parentId = getParentId(parentId);
+      }
+    }
+
+    return nextRow;
+  };
+
+  resetSelection();
```

Comment:
Is it a good idea to reset the selections on every shift click? Or is there a better idea? I am doing this reset (deselect any previous selections) since that is what File Explorer does.

Reply:
this makes sense to me

Reply:
if you maintain your own selection state (as mentioned in https://github.com/iTwin/iTwinUI-react/pull/774#discussion_r960020084) then i think you won't need to reset selection

Reply:
Thanks for the idea. With Vilius's code ideas from the `vilius/shift-select-demo`, we are now maintaining our own `Record/{}` of the selected rows and then setting `selectedRowIds: selectedRowIds` in the new state.

```tsx
const newState = {
  ...state,
  selectedRowIds: selectedRowIds,
};
```

This way works better than the toggle each row version since now `onSelect` is called just once (in the toggle version it was called on each toggle).

But the flickering problem is still unfortunately there. But I think that is not related to this PR. Since this flickering happens also on the hosted storybook.